projects/fal/src/fal/workflows.py

```diff
@@ -0,0 +1,426 @@
+            return input.execute(self)
+        else:
+            return input
+
+
+@dataclass
+class Leaf:
+    def execute(self, context: Context) -> JSONType:
+        raise NotImplementedError
+
+    def __getattr__(self, name: str) -> AttributeLeaf:
+        return AttributeLeaf(self, name)
+
+    def __getitem__(self, index: int) -> IndexLeaf:
+        return IndexLeaf(self, index)
+
+    @property
+    def referee(self) -> ReferenceLeaf:
+        raise NotImplementedError
+
+
+@dataclass
+class AttributeLeaf(Leaf):
+    leaf: Leaf
+    attribute: str
+
+    def execute(self, context: Context) -> JSONType:
+        output = self.leaf.execute(context)
+        assert isinstance(output, dict), f"{self.leaf!r} is not a dict"
+        return output[self.attribute]
+
+    def __repr__(self) -> str:
+        return f"{self.leaf!r}.{self.attribute}"
+
+    @property
+    def referee(self) -> ReferenceLeaf:
+        return self.leaf.referee
+
+
+@dataclass
+class IndexLeaf(Leaf):
+    leaf: Leaf
+    index: int
+
+    def execute(self, context: Context) -> JSONType:
+        output = self.leaf.execute(context)
+        assert isinstance(output, list), f"{self.leaf!r} is not an array"
+        return output[self.index]
+
+    def __repr__(self) -> str:
+        return f"{self.leaf!r}.{self.index}"
+
+    @property
+    def referee(self) -> ReferenceLeaf:
+        return self.leaf.referee
+
+
+@dataclass
+class ReferenceLeaf(Leaf):
+    id: str
+
+    def execute(self, context: Context) -> JSONType:
+        return context.vars[self.id]
+
+    def __repr__(self) -> str:
+        return VARIABLE_PREFIX + self.id
+
+    @property
+    def referee(self) -> ReferenceLeaf:
+        return self
+
+
+@dataclass
+class Node:
+    id: str
+
+    @classmethod
+    def from_json(cls, data: dict[str, Any]) -> Node:
+        type = data.pop("type")
+        if type == "display":
+            return Display.from_json(data)
+        elif type == "run":
+            return Run.from_json(data)
+        else:
+            raise ValueError(f"Invalid node type: {type}")
+
+    def to_json(self) -> dict[str, Any]:
+        raise NotImplementedError
+
+    def execute(self, context: Context) -> JSONType:
+        raise NotImplementedError
+
+    @property
+    def requires(self) -> set[str]:
+        raise NotImplementedError
+
+
+@dataclass
+class Display(Node):
+    fields: list[Leaf]
+
+    @classmethod
+    def from_json(cls, data: dict[str, Any]) -> Display:
+        return cls(
+            id=data["id"],
+            fields=import_workflow_json(data["fields"]),  # type: ignore
+        )
+
+    def to_json(self) -> dict[str, Any]:
+        return {
+            "type": "display",
+            "id": self.id,
+            "fields": export_workflow_json(self.fields),
+        }
+
+    def execute(self, context: Context) -> JSONType:
+        for url in context.hydrate(self.fields):  # type: ignore
+            webbrowser.open(url)
```

Comment:
Can we add a way to opt out of this? Maybe just print the url for people that don't want their browser to randomly open.